### PR TITLE
[Feature] Batch update partition's property in one command

### PR DESCRIPTION
**Pull request by xy720**
#3389 
We used to update single partition's properties by the following command:

```
alter table tbl_name modify partition p1 set ("replication_num" = "3");
```

Now this commit supports updating multiple partitions' properties of range table in one command.

Example:

- replication_num

```
alter table tbl_name modify partition (p1, p2, p3) set ("replication_num" = "3");
```

- storage_medium && storage_cooldown_time

```
alter table tbl_name modify partition (p1, p2, p3) set ("storage_medium" = "SSD", "storage_cooldown_time" = "2018-01-01 12:00:00");
```

- in_memory

```
alter table tbl_name modify partition (*) set ("in_memory" = "true");
```

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java [hunk 1]
```diff
@@ -48,10 +56,27 @@ public ModifyPartitionClause(String partitionName, Map<String, String> propertie
...
+    // c'tor for 'Modify Partition(*)' clause
+    private ModifyPartitionClause(Map<String, String> properties) {
+        super(AlterOpType.MODIFY_PARTITION);
+        this.partitionNames = Lists.newArrayList();
+        this.properties = properties;
+        this.needExpand = true;
+        this.needTableStable = false;
+    }
+
+    static public ModifyPartitionClause createStarClause(Map<String, String> properties) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java:68]:
```suggestion
    public static ModifyPartitionClause createStarClause(Map<String, String> properties) {
```

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 2]
```diff
@@ -3334,6 +3334,45 @@ public void replayRecoverPartition(RecoverInfo info) {
...
+                throw new DdlException(
+                        "Partition[" + partitionName + "] does not exist in table[" + olapTable.getName() + "]");
+            }
+        }
+
+        // If error occurs tell user which partition is wrong.
+        for (String partitionName : partitionNames) {
+            try {
+                Map<String, String> partitionProperties = Maps.newHashMap(properties);
+                modifyPartitionProperty(db, olapTable, partitionName, partitionProperties);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Catalog.java:3363]: better to write meta in one edit log

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java [hunk 3]
```diff
@@ -17,28 +17,36 @@
 
 package org.apache.doris.analysis;
 
+import com.clearspring.analytics.util.Lists;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java:20]: Wrong import.

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java [hunk 4]
```diff
@@ -64,11 +89,17 @@ public void analyze(Analyzer analyzer) throws AnalysisException {
...
+        return this.needExpand;
+    }
+
     @Override
     public String toSql() {
         StringBuilder sb = new StringBuilder();
         sb.append("MODIFY PARTITION ");
-        sb.append(partitionName);
+        sb.append("(");
+        sb.append(Joiner.on(", ").join(partitionNames));
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java:101]: Should print "*" if needExpand is true

**Review by morningman — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java (hunk 3 above)
Comment (xy720) [reply fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java]: fix

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java (hunk 4 above)
Comment (xy720) [reply fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java]: fixed

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java [hunk 5]
```diff
@@ -17,28 +17,41 @@
 
 package org.apache.doris.analysis;
 
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java:24]: Import order

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java [hunk 6]
```diff
@@ -48,27 +61,80 @@ public ModifyPartitionClause(String partitionName, Map<String, String> propertie
...
+        // check properties here
+        preCheck(Maps.newHashMap(properties));
+    }
+
+    // pre check the following properties' legality before modifying partition.
+    // 1. replication_num
+    // 2. storage_medium && storage_cooldown_time
+    // 3. in_memory
+    // 4. tablet type
+    public static void preCheck(Map<String, String> properties) throws AnalysisException {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java:100]: private, and no need to be static. And method name can be changed to `checkProperties`.

File: fe/src/main/java/org/apache/doris/persist/BatchModifyPartitionsInfo.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * used for batch modify multi partitions in one atomic operation
+ */
+
+public class BatchModifyPartitionsInfo implements Writable {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/persist/BatchModifyPartitionsInfo.java:34]: Add UT to test the read/write method of this class

File: fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java (hunk 6 above)
Comment (xy720) [reply fe/src/main/java/org/apache/doris/analysis/ModifyPartitionClause.java]: done

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**